Add 'on_disk_change' property to opennebula_virtual_machine

danhigham opened the pull request:
Add a property called 'on_disk_change' to the virtual machine resource;

```
  resource "opennebula_virtual_machine" "test_vm" {
     count       = 1
     name        = "test-vm"
     permissions = "660"
     cpu        = 1
     vcpu = 1
     memory     = 1024

     ...

     on_disk_change = "recreate"
  }
```

on_disk_change (if the disks image id has changed) can have three different values;

"recreate" - recreate the vm (call ForceNew)
"swap" - just swap the disk as normal

This pull request is in response to https://github.com/OpenNebula/terraform-provider-opennebula/issues/156

treywelsh commented on opennebula/resource_opennebula_virtual_machine.go:
```diff
@@ -126,6 +126,18 @@ func resourceOpennebulaVirtualMachine() *schema.Resource {
+				Default:  "swap", //"recreate", "restart" or "swap",
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					value := v.(string)
+					if value != "recreate" && value != "restart" && value != "swap" {
```
It's not a problem, but, to be consistent with the code already in the provider the values ("recreate", "restart" or "swap") could be stored in a slice, and you could test for `value` via the function `inArray(value, sliceOfPossibleValues)`.
Look at some other ValidateFunc methods

danhigham replied:
Done :-)

treywelsh commented on opennebula/resource_opennebula_virtual_machine.go:
```diff
@@ -1125,6 +1137,25 @@ func resourceOpennebulaVirtualMachineUpdate(d *schema.ResourceData, meta interfa
+
+			if err != nil {
+				return fmt.Errorf(
+					"Error waiting for virtual machine (ID:%d) to be in state %s: %s", vmc.ID, strings.Join(vmDiskUpdateReadyStates, " "), err)
```
I think that `strings.Join(vmDiskUpdateReadyStates, " ")` is a copy/paste of an error I made, just replace it by  `"RUNNING"`: it's the only state you are waiting for

danhigham replied:
Done.

treywelsh commented on opennebula/resource_opennebula_virtual_machine.go:
```diff
@@ -1504,6 +1535,27 @@ func generateVm(d *schema.ResourceData, tplContext *dyn.Vector) (string, error)
+
+	for i, oi := range oldDiskList {
+		oiB := oi.(map[string]interface{})
+		niB := newDiskList[i].(map[string]interface{})
```
If the user remove a disk newDiskList will be shorter of 1 element, and `i` will point outside of the slice. To avoid this you'll have to check the slice length before the loop.
Also consider the user adding a new disk, it will be ignored

treywelsh commented on opennebula/resource_opennebula_virtual_machine.go:
```diff
@@ -1504,6 +1535,27 @@ func generateVm(d *schema.ResourceData, tplContext *dyn.Vector) (string, error)
+		log.Printf("[DEBUG] new disk: %s\n", niB["image_id"])
+
+		if oiB["image_id"].(int) != niB["image_id"].(int) {
+			if onChange == "recreate" {
```
instead of only test this line, we could put this whole part of code in the test (this would save a bit of useless computation in cases onChange != "recreate")

danhigham replied:
Done.

treywelsh commented on opennebula/resource_opennebula_virtual_machine.go:
```diff
@@ -1504,6 +1535,27 @@ func generateVm(d *schema.ResourceData, tplContext *dyn.Vector) (string, error)
+		log.Printf("[DEBUG] old disk: %s\n", oiB["image_id"])
+		log.Printf("[DEBUG] new disk: %s\n", niB["image_id"])
+
+		if oiB["image_id"].(int) != niB["image_id"].(int) {
```
The full behaviour around disk changes wasn't defined, among other we could have some corner cases. Are they acceptable for everyone (feel free to discuss) ?

For instance: 
- Even if the user add a new disk (other than the OS disk like described in the issue), this will trigger the behavior (recreate/restart).
- If an user swap two disk image IDs other than the OS disk (and potentially update some other fields  of the disks): from A B C to A C B
  we'll have almost have the same disk list, but this change will make the provider recreate/restart the VM

danhigham replied:
The original intention was to modify behavior for when an attached disk was changed, not added or removed. If you have other behavioral suggestions, I will happily implement them too.

treywelsh replied:
I'm ok with the behavior, however it remains some work. Below I'll try to be clear and as accurate as I can, so sorry for the bunch of text.

I consider a disk as "changed" when some disk attributes changed, so compare only the image_id of the disks is not sufficient.
If other attributes than image_id changed (target, driver...) this will trigger a detachment followed by an attachment of the disk (AFAIK there is no update-in-place method for disk in OpenNebula, except resize method).
So, we want to find the disks that have the same image_id with some attributes changed to know if we should trigger the on_disk_change behavior.
For now I'll ignore the case when an image would be attached twice to the same VM (two disk with same image_id, introduced by #125).

They is several ways to implement this: from scratch with a double loop to compare the list of disks and find disks changed, or inspired on how I managed disk update, or a better idea if you have one.

Here is how I managed disk update:
I added an helper to retrieve the lists of disks to attach/detach:
https://github.com/OpenNebula/terraform-provider-opennebula/search?q=diffListConfig
As you can see, I compare some sets of disks based on attributes image_id, target, driver (changes on other attributes would be ignored).
The helper returns us toAttach and toDetach lists, but they include more disks that we want: toAttach contains new disks to attach (new image_id introduced), and toDetach contains disk that we only want to detach.
At least, there is some additional filtering to do on these list to determine if there is some disks changed.

I hope this help, it's nice to have other externals contributors so feel free to drop some comments if I can help more, and if you have a better idea, please share it

treywelsh replied:
@danhigham about this comment ?

danhigham replied:
Apologies for the delay on this, I will work on getting these changes in this week.

treywelsh reviewed (commented):
Many thanks for your contribution, I think that's a good starting point for this feature.

It still miss some parts like documentation, changelog update, and we may need a some more discussions to clearly specify the behavior

danhigham commented:
@treywelsh You mentioned that the PR may be missing documentation and a change-log update. Can you provide starting points for both so I can work toward getting those done.

treywelsh commented:
To work on up to date code, don't forget to rebase your branch

danhigham commented:
@treywelsh Done :-)

danhigham commented:
Rebased this with current master

treywelsh commented:
Hi,

It seems that the way you updated your PRs via multiple merges scramble the diff on github, it also adds some master changes to the diff and we can't review it from github.
If you look at the PR details, this one show 41 commits.
Looking at git log with --graph, this show a lot of branches.

I don't pretend to be a git expert but this seems a bit complex so please, could you make a branch that contains just your commits on top of master without the mutiple merges ?

danhigham commented:
Hi, I will do this today. This _was_ originally how I had done it but then the PR was left for so long I figured you would prefer it rebase on the current master, but obviously that process complicates things in that respect.

danhigham commented:
@treywelsh Ok, I squashed all the changes to one commit and re-pushed to the branch the PR is based on.

jaypif commented on opennebula/resource_opennebula_virtual_machine.go:
```diff
@@ -1189,6 +1202,25 @@ func resourceOpennebulaVirtualMachineUpdate(d *schema.ResourceData, meta interfa
 		}
+
+		// If the value of "on_disk_change" is "restart", start the vm
+		if strings.ToUpper(d.Get("on_disk_change").(string)) == "RESTART" {
```
I am not sure it's a good idea to propose a "restart" option for disk change.

IMHO, The provider and usually Terraform is used to managed infrastructure not, VM states.

In you case even if the infrastructure is correctly changed, if your VM is not in "RUNNNING" state for an external reason, Terraform will consider there is an error while it s correct in terms of Infrastructure.

I agree about `swap` or `recreate `option` by the way.

Open to discuss about the restart option.

danhigham replied:
More than happy to remove the restart option. I agree with your point.

danhigham commented:
@jaypif @treywelsh Ok, I removed the restart option and changed the resourceVMCustomizeDiff function for VMs to use the diffListConfig helper method @treywelsh mentioned.

danhigham commented:
@treywelsh I notice the CI-6_2 pipeline is failing although this seems to be unrelated to my changes, is that assumption correct?

treywelsh commented:
It's correct, here is the related issue: #178

danhigham commented:
> It's correct, here is the related issue: #178

Great, is there anything else I can provide to help get this PR across the finish line?

treywelsh commented:
I'm testing the PR a bit to be sure it works in some basic cases.

treywelsh commented on opennebula/resource_opennebula_virtual_machine.go:
```diff
@@ -1712,6 +1725,29 @@ func generateVm(d *schema.ResourceData, tplContext *dyn.Vector) (string, error)
+			"volatile_format")
+
+		if len(toDetach) > 0 {
+			diff.ForceNew(fmt.Sprintf("disk.%d.image_id", toDetach[0]))
```
It works on an `image_id` change, not for other attributes changes, and I'm not sure that it will works in some cases with several disks attached.
If we want the behavior to be triggered only on `image_id` change, we should write this instead:
```
		toDetach, toAttach := diffListConfig(newDiskList, oldDiskList,
			&schema.Resource{
				Schema: diskFields(),
			},
			"image_id")
```

Now, here is the problem: I'm not sure for the key in `diff.ForceNew`.
The `%d` in `disk.%d.image_id` should be the disk position in the list. We want to enforce the `ForceNew` behavior when a given attribute of the disk has changed.
The problem is that with the `diffListConfig` method we lose the disk position in the lists returned.

I ended up with this brute forcing code, I'm not fully satisfied but this seems to work:
```
	if strings.ToUpper(onChange) == "RECREATE" {

		oldDisk, newDisk := diff.GetChange("disk")
		oldDiskList := oldDisk.([]interface{})
		newDiskList := newDisk.([]interface{})
		toDetach, toAttach := diffListConfig(newDiskList, oldDiskList,
			&schema.Resource{
				Schema: diskFields(),
			},
			"image_id",
			"target",
			"driver")

		if len(toDetach) > 0 {
			for i := range oldDiskList {
				diff.ForceNew(fmt.Sprintf("disk.%d.image_id", i))
				diff.ForceNew(fmt.Sprintf("disk.%d.target", i))
				diff.ForceNew(fmt.Sprintf("disk.%d.driver", i))
			}
		}
```
If we are in the case where we want to detach a disk, this iterate over all disks and call ForceNew on some attributes. If nothing has changed, the ForceNew method returns an error (that I shamefully ignore), else, this generate a new diff that show which field changed and triggered the VM recreation.

There is probaly some other solutions, maybe simpler, feel free to test this with several disks or to propose something better

danhigham replied:
@treywelsh I somehow missed this comment, thinking this PR was ready, as @jaypif had already approved the change. Does diffListConfig not show which property had changed?

danhigham replied:
@treywelsh I went ahead and went with your suggestion, it seems to work just fine. Do you think we can merge this soon?

treywelsh replied:
Sorry for this, I made the test just after he reads the PR.

I won't work on the provider for one week, then we just need to focus on 1 or 2 bug fixes to release the 0.4.2 version (this should be quick) then I'll take a look at your PR, test it and it should be good.

This will be part of the 0.5.0 but there is still some work to do to release it (we plan a bunch of things then we'll need to test it a bit more in depth etc.).

I suppose you need this quickly ?

danhigham replied:
@treywelsh It's going to greatly increase the speed of a deployment process we have, but it's all good. It looks like you're releasing on a cadence of every month or so. In the grand scheme of things, i've been working on it with you since November, so another month or so won't hurt!

treywelsh commented:
By the way, we are close to something that works. Overall it's a good work so thank you for the time you spend on this PR

danhigham committed:
Re-base feature with latest from master branch

danhigham commented:
@treywelsh Just re-based this with the latest from master. Is this looking likely to make it to the next release?

treywelsh commented:
At least it miss a line in the CHANGELOG to describe your changes.

Appart from this the PR seems ok to me. I made a quick test and this work.
Did you made some tests on your side to be sure it works the way you expect ?

If everyone is happy with this I'll merge it soon to make it available in the 0.5.0 release.

danhigham committed:
Add entry to changelog for on_disk_change property

danhigham commented:
@treywelsh We're happy with it, I added a feature line to the changelog.
